### PR TITLE
#1275 Fix for GROUPBY with multivalued fields where not all fields have every sub-field populated.

### LINKED ISSUE #1275: GroupBy needs to handle multivalued fields from atoms where not every field is populated in every atom.

### PR DESCRIPTION
**Pull request by friedlou**
Fixes #1275

### CODE REVIEW

File: warehouse/query-core/src/test/java/datawave/query/util/VisibilityWiseGuysIngest.java [hunk 1]
```diff
@@ -82,9 +82,17 @@ public static void writeItAll(Connector con, WhatKindaRange range) throws Except
             mutation.put(datatype + "\u0000" + corleoneUID, "BIRTHDAY.3" + "\u0000" + "4", columnVisibilityItalian, timeStamp, emptyValue);
             mutation.put(datatype + "\u0000" + corleoneUID, "BIRTHDAY.4" + "\u0000" + "5", columnVisibilityItalian, timeStamp, emptyValue);
             mutation.put(datatype + "\u0000" + corleoneUID, "BIRTHDAY.5" + "\u0000" + "22", columnVisibilityItalian, timeStamp, emptyValue);
+            // to test whether sparse fields are properly ignored while grouping, leave dependents field blank for group 0 and 3
+            // mutation.put(datatype + "\u0000" + corleoneUID, "DEPENDENTS.0" + "\u0000" + "2", columnVisibilityItalian, timeStamp, emptyValue);
```
Comment (hlgp) [inline warehouse/query-core/src/test/java/datawave/query/util/VisibilityWiseGuysIngest.java]: if you don't need this commented out code, please remove

Comment (friedlou) [reply warehouse/query-core/src/test/java/datawave/query/util/VisibilityWiseGuysIngest.java]: Done

File: warehouse/query-core/src/main/java/datawave/query/common/grouping/GroupingUtil.java [hunk 2]
```diff
@@ -90,18 +90,18 @@ public GroupingInfo getGroupingInfo(Map.Entry<Key,Document> entry, Set<String> g
...
                     if (gtNames == null || gtNames.isEmpty()) {
                         log.trace("gtNames: {}", gtNames);
                         log.trace("fieldToFieldWithContextMap: {} did not contain: {}", fieldToFieldWithContextMap, fieldListItem);
                     } else {
-                        String gtName = gtNames.iterator().next();
-                        int idx = gtName.indexOf('.');
-                        if (idx != -1) {
-                            currentGroupingContext = gtName.substring(idx + 1);
+                        String nameWithGrouping = fieldListItem + "." + currentGroupingContext;
+                        final String gtName = gtNames.stream().filter(name -> nameWithGrouping.equals(name)).findAny().orElse(null);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/common/grouping/GroupingUtil.java]: The fieldToFieldWithContextMap is being created as a TreeMultimap which returns a SortedSet (in fact a NavigatableSet) on the get call.  So gtNames is actually a NavigatableSet which means that getNames.get(name) is all you need here!  Doing a stream can be extremely costly when the number of groups is very large (which is definitely the case sometimes).
If you want to make this more obvious, change the getFieldToFieldWithGroupingContextMap routine to return a TreeMultimap instead, change fieldToFieldWithContextMap to be a TreeMultimap as well and change gtNames to be a NavigatableSet.

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/common/grouping/GroupingUtil.java (hunk 2 above)
Comment (friedlou) [reply warehouse/query-core/src/main/java/datawave/query/common/grouping/GroupingUtil.java]: Please look a the latest commit.

**Review by ivakegg — APPROVED** [review]

**Review by hlgp — APPROVED** [review]

**Pull request merged**